spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -176,88 +219,189 @@ function verifyChannelUpgradeErrorAbsence(
+
+    // get last packet sent on channel and set it in the upgrade struct
+    // last packet sent is the nextSequenceSend on the channel minus 1
+    lastPacketSendSequence = publicStore.get(nextSequenceSendPath(portIdentifier, channelIdentifier)) - 1
+    upgrade = Upgrade{
+        fields: proposedFields,
+        timeout: timeout,
+        lastPacketSent: lastPacketSendSequence,
+    }
+
+    // store upgrade in public store for counterparty proof verification
+    publicStore.set(channelUpgradePath(portIdentifier, channelIdentifier), upgrade)
+
+    currentChannel.sequence = currentChannel.sequence + 1
+    currentChannel.state = INITUPGRADE
+    publicStore.set(channelPath(portIdentifier, channelIdentifier), channel)
+    return currentChannel.sequence
 }
 ```
 
-The upgrade timeout path MUST have an associated verification membership method on the connection interface in order for a counterparty to prove that a chain stored a particular timeout in the upgrade timeout path.
+`startFlushUpgradeHandshake` will set the counterparty last packet send and continue blocking the upgrade from continuing until all in-flight packets have been flushed. When the channel is in blocked mode, any packet receive above the counterparty last packet send will be rejected. It will verify the upgrade parameters and set the channel state to one of the flushing states (`TRYUPGRADE` or `ACKUPGRADE`) passed in by caller, set the `FlushStatus` to `FLUSHING` and block sendpackets. During this time; `receivePacket`, `acknowledgePacket` and `timeoutPacket` will still be allowed and processed according to the original channel parameters. The new proposed upgrade will be stored in the public store for counterparty verification.
 
 ```typescript
-// Connection VerifyChannelUpgradeTimeout method
-function verifyChannelUpgradeTimeout(
-  connection: ConnectionEnd,
-  height: Height,
-  proof: CommitmentProof,
-  counterpartyPortIdentifier: Identifier,
-  counterpartyChannelIdentifier: Identifier,
-  upgradeTimeout: UpgradeTimeout, 
+function startFlushUpgradeHandshake(
+    portIdentifier: Identifier,
+    channelIdentifier: Identifier,
+    proposedUpgradeFields: UpgradeFields,
+    counterpartyChannel: ChannelEnd,
+    counterpartyUpgrade: Upgrade,
+    channelState: ChannelState,
+    proofChannel: CommitmentProof,
+    proofUpgrade: CommitmentProof,
+    proofHeight: Height
 ) {
-    clientState = queryClientState(connection.clientIdentifier)
-    path = applyPrefix(connection.counterpartyPrefix, channelUpgradeTimeoutPath(counterpartyPortIdentifier, counterpartyChannelIdentifier))
-    return verifyMembership(clientState, height, 0, 0, proof, path, upgradeTimeout)
+    abortTransactionUnless(channelState == TRYUPGRADE || channelState == ACKUPGRADE)
+
+    // get underlying connection for proof verification
+    connection = getConnection(currentChannel.connectionIdentifier)
+    counterpartyHops = getCounterpartyHops(connection)
```

Comment:
why do we need this? we are receiving the `counterpartyChannel` already as a param